File: OsmAnd-java/src/main/java/net/osmand/router/RouteColorize.java
package net.osmand.router;

import net.osmand.ColorPalette;
import net.osmand.PlatformUtil;
import net.osmand.gpx.GPXFile;
import net.osmand.gpx.GPXTrackAnalysis;
import net.osmand.gpx.GPXUtilities.Track;
import net.osmand.gpx.GPXUtilities.TrkSegment;
import net.osmand.gpx.GPXUtilities.WptPt;
import net.osmand.osm.edit.Node;
import net.osmand.osm.edit.OsmMapUtils;
import net.osmand.util.Algorithms;
import net.osmand.util.MapUtils;

import org.apache.commons.logging.Log;

import java.util.ArrayList;
import java.util.List;

@Deprecated
public class RouteColorize {

	public static double MAX_CORRECT_ELEVATION_DISTANCE = 100.0;// in meters
	public static int SLOPE_RANGE = 150;// 150 meters
	private static final Log LOG = PlatformUtil.getLog(RouteColorize.class);
	private static final float DEFAULT_BASE = 17.2f;
	private static final double MIN_DIFFERENCE_SLOPE = 0.05d;// 5%


	private double[] latitudes;
	private double[] longitudes;
	private double[] values;
	private double minValue;
	private double maxValue;
	private ColorPalette palette;
	private List<RouteColorizationPoint> dataList;
	private ColorizationType colorizationType;

	public enum ColorizationType {
		ELEVATION, SPEED, SLOPE, NONE
	}

	/**
	 * @param minValue can be NaN
	 * @param maxValue can be NaN
	 * @param palette  array {{value,color},...} - color in sRGB (decimal) format OR
	 *                 {{value,RED,GREEN,BLUE,ALPHA},...} - color in RGBA format
	 */
	public RouteColorize(double[] latitudes, double[] longitudes, double[] values, double minValue, double maxValue,
			ColorPalette palette) {
		this.latitudes = latitudes;
		this.longitudes = longitudes;
		this.values = values;
		this.minValue = minValue;
		this.maxValue = maxValue;
		this.palette = palette;
		if (Double.isNaN(minValue) || Double.isNaN(maxValue)) {
			calculateMinMaxValue();
		}
		if (palette == null || palette.getColors().size() < 2) {
			this.palette = new ColorPalette(ColorPalette.MIN_MAX_PALETTE, minValue, maxValue);
		} else {
			this.palette = palette;
		}
	}

	/**
	 * @param type ELEVATION, SPEED, SLOPE
	 */
	public RouteColorize(GPXFile gpxFile, ColorizationType type, ColorPalette palette) {
		this(gpxFile, null, type, palette, 0);
	}

	public RouteColorize(GPXFile gpxFile, GPXTrackAnalysis analysis, ColorizationType type,
			ColorPalette palette, float maxProfileSpeed) {
		if (!gpxFile.hasTrkPt()) {
			LOG.warn("GPX file is not consist of track points");
			return;
		}
		List<Double> latList = new ArrayList<>();
		List<Double> lonList = new ArrayList<>();
		List<Double> valList = new ArrayList<>();
		int wptIdx = 0;
		if (analysis == null) {
			long time = Algorithms.isEmpty(gpxFile.path) ? System.currentTimeMillis() : gpxFile.modifiedTime;
			analysis = gpxFile.getAnalysis(time);
		}
		for (Track t : gpxFile.tracks) {
			for (TrkSegment ts : t.segments) {
				if (ts.generalSegment || ts.points.size() < 2) {
					continue;
				}

				for (WptPt p : ts.points) {
					latList.add(p.lat);
					lonList.add(p.lon);
					if (type == ColorizationType.SPEED) {
						valList.add((double) analysis.pointAttributes.get(wptIdx).speed);
					} else {
						valList.add((double) analysis.pointAttributes.get(wptIdx).elevation);
					}
					wptIdx++;
				}
			}
		}

		colorizationType = type;
		latitudes = listToArray(latList);
		longitudes = listToArray(lonList);

		if (type == ColorizationType.SLOPE) {
			values = calculateSlopesByElevations(latitudes, longitudes, listToArray(valList), SLOPE_RANGE);
		} else {
			values = listToArray(valList);
		}
		calculateMinMaxValue(analysis, maxProfileSpeed);
		if (type == ColorizationType.SLOPE) {
			this.palette = isValidPalette(palette) ? palette : ColorPalette.SLOPE_PALETTE;
		} else {
			this.palette = new ColorPalette(isValidPalette(palette) ? palette : ColorPalette.MIN_MAX_PALETTE, minValue,
					maxValue);
		}
	}

	private boolean isValidPalette(ColorPalette palette) {
		return palette != null && palette.getColors().size() >= 2;
	}

	/**
	 * Calculate slopes from elevations needs for right colorizing
	 *
	 * @param slopeRange - in what range calculate the derivative, usually we used
	 *                   150 meters
	 * @return slopes array, in the begin and the end present NaN values!
	 */
	public double[] calculateSlopesByElevations(double[] latitudes, double[] longitudes, double[] elevations,
			double slopeRange) {
		correctElevations(latitudes, longitudes, elevations);
		double[] newElevations = elevations;
		for (int i = 2; i < elevations.length - 2; i++) {
			newElevations[i] = elevations[i - 2] + elevations[i - 1] + elevations[i] + elevations[i + 1]
					+ elevations[i + 2];
			newElevations[i] /= 5;
		}
		elevations = newElevations;

		double[] slopes = new double[elevations.length];
		if (latitudes.length != longitudes.length || latitudes.length != elevations.length) {
			LOG.warn("Sizes of arrays latitudes, longitudes and values are not match");
			return slopes;
		}

		double[] distances = new double[elevations.length];
		double totalDistance = 0.0d;
		distances[0] = totalDistance;
		for (int i = 0; i < elevations.length - 1; i++) {
			totalDistance += MapUtils.getDistance(latitudes[i], longitudes[i], latitudes[i + 1], longitudes[i + 1]);
			distances[i + 1] = totalDistance;
		}

		for (int i = 0; i < elevations.length; i++) {
			if (distances[i] < slopeRange / 2 || distances[i] > totalDistance - slopeRange / 2) {
				slopes[i] = Double.NaN;
			} else {
				double[] arg = findDerivativeArguments(distances, elevations, i, slopeRange);
				slopes[i] = (arg[1] - arg[0]) / (arg[3] - arg[2]);
			}
		}
		return slopes;
	}

	private void correctElevations(double[] latitudes, double[] longitudes, double[] elevations) {
		for (int i = 0; i < elevations.length; i++) {
			if (Double.isNaN(elevations[i])) {
				double leftDist = MAX_CORRECT_ELEVATION_DISTANCE;
				double rightDist = MAX_CORRECT_ELEVATION_DISTANCE;
				double leftElevation = Double.NaN;
				double rightElevation = Double.NaN;
				for (int left = i - 1; left > 0 && leftDist <= MAX_CORRECT_ELEVATION_DISTANCE; left--) {
					if (!Double.isNaN(elevations[left])) {
						double dist = MapUtils.getDistance(latitudes[left], longitudes[left], latitudes[i],
								longitudes[i]);
						if (dist < leftDist) {
							leftDist = dist;
							leftElevation = elevations[left];
						} else {
							break;
						}
					}
				}
				for (int right = i + 1; right < elevations.length
						&& rightDist <= MAX_CORRECT_ELEVATION_DISTANCE; right++) {
					if (!Double.isNaN(elevations[right])) {
						double dist = MapUtils.getDistance(latitudes[right], longitudes[right], latitudes[i],
								longitudes[i]);
						if (dist < rightDist) {
							rightElevation = elevations[right];
							rightDist = dist;
						} else {
							break;
						}
					}
				}
				if (!Double.isNaN(leftElevation) && !Double.isNaN(rightElevation)) {
					elevations[i] = (leftElevation + rightElevation) / 2;
				} else if (Double.isNaN(leftElevation) && !Double.isNaN(rightElevation)) {
					elevations[i] = rightElevation;
				} else if (!Double.isNaN(leftElevation) && Double.isNaN(rightElevation)) {
					elevations[i] = leftElevation;
				} else {
					for (int right = i + 1; right < elevations.length; right++) {
						if (!Double.isNaN(elevations[right])) {
							elevations[i] = elevations[right];
							break;
						}
					}
				}
			}
		}
	}

	public List<RouteColorizationPoint> getResult() {
		List<RouteColorizationPoint> result = new ArrayList<>();
		for (int i = 0; i < latitudes.length; i++) {
			result.add(new RouteColorizationPoint(i, latitudes[i], longitudes[i], values[i]));
		}
		setColorsToPoints(result);
		return result;
	}

	public List<RouteColorizationPoint> getSimplifiedResult(int simplificationZoom) {
		List<RouteColorizationPoint> simplifiedResult = simplify(simplificationZoom);
		setColorsToPoints(simplifiedResult);
		return simplifiedResult;
	}

	private void setColorsToPoints(List<RouteColorizationPoint> points) {
		for (RouteColorizationPoint point : points) {
			point.primaryColor = palette.getColorByValue(point.val);
		}
	}

	public void setPalette(ColorPalette palette) {
		if (palette != null) {
			this.palette = palette;
		}
	}

	public List<RouteColorizationPoint> simplify(int simplificationZoom) {
		if (dataList == null) {
			dataList = new ArrayList<>();
			for (int i = 0; i < latitudes.length; i++) {
				dataList.add(new RouteColorizationPoint(i, latitudes[i], longitudes[i], values[i]));
			}
		}
		List<Node> nodes = new ArrayList<>();
		List<Node> result = new ArrayList<>();
		for (RouteColorizationPoint data : dataList) {
			nodes.add(new net.osmand.osm.edit.Node(data.lat, data.lon, data.id));
		}

		double epsilon = Math.pow(2.0, DEFAULT_BASE - simplificationZoom);
		result.add(nodes.get(0));
		OsmMapUtils.simplifyDouglasPeucker(nodes, 0, nodes.size() - 1, result, epsilon);

		List<RouteColorizationPoint> simplified = new ArrayList<>();
		for (int i = 1; i < result.size(); i++) {
			int prevId = (int) result.get(i - 1).getId();
			int currentId = (int) result.get(i).getId();
			List<RouteColorizationPoint> sublist = dataList.subList(prevId, currentId);
			simplified.addAll(getExtremums(sublist));
		}
		Node lastSurvivedPoint = result.get(result.size() - 1);
		simplified.add(dataList.get((int) lastSurvivedPoint.getId()));
		return simplified;
	}

	private List<RouteColorizationPoint> getExtremums(List<RouteColorizationPoint> subDataList) {
		if (subDataList.size() <= 2) {
			return subDataList;
		}

		List<RouteColorizationPoint> result = new ArrayList<>();
		double min;
		double max;
		min = max = subDataList.get(0).val;
		for (RouteColorizationPoint pt : subDataList) {
			if (min > pt.val) {
				min = pt.val;
			}
			if (max < pt.val) {
				max = pt.val;
			}
		}

		double diff = max - min;

		result.add(subDataList.get(0));
		for (int i = 1; i < subDataList.size() - 1; i++) {
			double prev = subDataList.get(i - 1).val;
			double current = subDataList.get(i).val;
			double next = subDataList.get(i + 1).val;
			RouteColorizationPoint currentData = subDataList.get(i);

			if ((current > prev && current > next) || (current < prev && current < next)
					|| (current < prev && current == next) || (current == prev && current < next)
					|| (current > prev && current == next) || (current == prev && current > next)) {
				RouteColorizationPoint prevInResult;
				if (result.size() > 0) {
					prevInResult = result.get(0);
					if (prevInResult.val / diff > MIN_DIFFERENCE_SLOPE) {
						result.add(currentData);
					}
				} else
					result.add(currentData);
			}
		}
		result.add(subDataList.get(subDataList.size() - 1));
		return result;
	}


	/**
	 * @return double[minElevation, maxElevation, minDist, maxDist]
	 */
	private double[] findDerivativeArguments(double[] distances, double[] elevations, int index, double slopeRange) {
		double[] result = new double[4];
		double minDist = distances[index] - slopeRange / 2;
		double maxDist = distances[index] + slopeRange / 2;
		result[0] = Double.NaN;
		result[1] = Double.NaN;
		result[2] = minDist;
		result[3] = maxDist;
		int closestMaxIndex = -1;
		int closestMinIndex = -1;
		for (int i = index; i < distances.length; i++) {
			if (distances[i] == maxDist) {
				result[1] = elevations[i];
				break;
			}
			if (distances[i] > maxDist) {
				closestMaxIndex = i;
				break;
			}
		}
		for (int i = index; i >= 0; i--) {
			if (distances[i] == minDist) {
				result[0] = elevations[i];
				break;
			}
			if (distances[i] < minDist) {
				closestMinIndex = i;
				break;
			}
		}
		if (closestMaxIndex > 0) {
			double diff = distances[closestMaxIndex] - distances[closestMaxIndex - 1];
			double coef = (maxDist - distances[closestMaxIndex - 1]) / diff;
			if (coef > 1 || coef < 0) {
				LOG.warn("Coefficient fo max must be 0..1 , coef=" + coef);
			}
			result[1] = (1 - coef) * elevations[closestMaxIndex - 1] + coef * elevations[closestMaxIndex];
		}
		if (closestMinIndex >= 0) {
			double diff = distances[closestMinIndex + 1] - distances[closestMinIndex];
			double coef = (minDist - distances[closestMinIndex]) / diff;
			if (coef > 1 || coef < 0) {
				LOG.warn("Coefficient for min must be 0..1 , coef=" + coef);
			}
			result[0] = (1 - coef) * elevations[closestMinIndex] + coef * elevations[closestMinIndex + 1];
		}
		if (Double.isNaN(result[0]) || Double.isNaN(result[1])) {
			LOG.warn("Elevations wasn't calculated");
		}
		return result;
	}

	public static double getMinValue(ColorizationType type, GPXTrackAnalysis analysis) {
		switch (type) {
		case SPEED:
			return 0.0;
		case ELEVATION:
			return analysis.getMinElevation();
		case SLOPE:
			return ColorPalette.SLOPE_MIN_VALUE;
		default:
			return -1;
		}
	}

	public static double getMaxValue(ColorizationType type, GPXTrackAnalysis analysis, double minValue,
			double maxProfileSpeed) {
		switch (type) {
		case SPEED:
			return Math.max(analysis.getMaxSpeed(), maxProfileSpeed);
		case ELEVATION:
			return Math.max(analysis.getMaxElevation(), minValue + 50);
		case SLOPE:
			return ColorPalette.SLOPE_MAX_VALUE;
		default:
			return -1;
		}
	}


	private void calculateMinMaxValue() {
		if (values.length == 0)
			return;
		minValue = maxValue = Double.NaN;
		for (double value : values) {
			if ((Double.isNaN(maxValue) || Double.isNaN(minValue)) && !Double.isNaN(value))
				maxValue = minValue = value;
			if (minValue > value)
				minValue = value;
			if (maxValue < value)
				maxValue = value;
		}
	}

	private void calculateMinMaxValue(GPXTrackAnalysis analysis, float maxProfileSpeed) {
		calculateMinMaxValue();
		// set strict limitations for maxValue
		maxValue = getMaxValue(colorizationType, analysis, minValue, maxProfileSpeed);
	}

	private double[] listToArray(List<Double> doubleList) {
		double[] result = new double[doubleList.size()];
		for (int i = 0; i < doubleList.size(); i++) {
			result[i] = doubleList.get(i);
		}
		return result;
	}

	public static ColorPalette getDefaultPalette(ColorizationType colorizationType) {
		if (colorizationType == ColorizationType.SLOPE) {
			return ColorPalette.SLOPE_PALETTE;
		} else {
			return ColorPalette.MIN_MAX_PALETTE;
		}
	}

	public static class RouteColorizationPoint {

		public final int id;
		public final double lat;
		public final double lon;
		public final double val;

		public int primaryColor;
		public int secondaryColor;

		public RouteColorizationPoint(int id, double lat, double lon, double val) {
			this.id = id;
			this.lat = lat;
			this.lon = lon;
			this.val = val;
		}
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/routing/RouteColorize.kt
package net.osmand.shared.routing

import net.osmand.shared.ColorPalette
import net.osmand.shared.extensions.currentTimeMillis
import net.osmand.shared.gpx.GpxFile
import net.osmand.shared.gpx.GpxTrackAnalysis
import net.osmand.shared.util.KAlgorithms
import net.osmand.shared.util.KMapUtils
import net.osmand.shared.util.LoggerFactory
import kotlin.math.*

class RouteColorize {
	private lateinit var latitudes: DoubleArray
	private lateinit var longitudes: DoubleArray
	private lateinit var values: DoubleArray
	private lateinit var palette: ColorPalette
	private var fixedValues = false
	private var minValue = 0.0
	private var maxValue = 0.0
	private var dataList: MutableList<RouteColorizationPoint>? = null
	private var colorizationType: ColorizationType? = null

	enum class ColorizationType(
		val bipolar: Boolean = false
	) {
		ELEVATION,
		SPEED,
		SLOPE(bipolar = true),
		NONE;
	}

	/**
	 * @param minValue can be NaN
	 * @param maxValue can be NaN
	 * @param palette  array {{value,color},...} - color in sRGB (decimal) format OR
	 * {{value,RED,GREEN,BLUE,ALPHA},...} - color in RGBA format
	 */
	constructor(
		latitudes: DoubleArray,
		longitudes: DoubleArray,
		values: DoubleArray,
		minValue: Double,
		maxValue: Double,
		palette: ColorPalette?,
		fixedValues: Boolean
	) {
		this.latitudes = latitudes
		this.longitudes = longitudes
		this.values = values
		this.minValue = minValue
		this.maxValue = maxValue
		this.fixedValues = fixedValues
		if (minValue.isNaN() || maxValue.isNaN()) {
			calculateMinMaxValue()
		}
		if (palette == null || palette.colors.size < 2) {
			this.palette = ColorPalette(ColorPalette.MIN_MAX_PALETTE, minValue, maxValue)
		} else {
			this.palette = palette
		}
	}

	/**
	 * @param type ELEVATION, SPEED, SLOPE
	 */
	constructor(
		gpxFile: GpxFile,
		type: ColorizationType,
		palette: ColorPalette?,
		fixedValues: Boolean
	) : this(gpxFile, null, type, palette, 0f, fixedValues)

	constructor(
		gpxFile: GpxFile,
		analysis: GpxTrackAnalysis?,
		type: ColorizationType,
		palette: ColorPalette?,
		maxProfileSpeed: Float,
		fixedValues: Boolean
	) {
		this.fixedValues = fixedValues
		var analysis: GpxTrackAnalysis? = analysis
		if (!gpxFile.hasTrkPt()) {
			LOG.warn("GPX file is not consist of track points")
			return
		}
		val latList = mutableListOf<Double>()
		val lonList = mutableListOf<Double>()
		val valList = mutableListOf<Double>()
		var wptIdx = 0
		if (analysis == null) {
			val time: Long =
				if (KAlgorithms.isEmpty(gpxFile.path)) currentTimeMillis() else gpxFile.modifiedTime
			analysis = gpxFile.getAnalysis(time)
		}
		for (t in gpxFile.tracks) {
			for (ts in t.segments) {
				if (ts.generalSegment || ts.points.size < 2) {
					continue
				}
				for (p in ts.points) {
					latList.add(p.lat)
					lonList.add(p.lon)
					if (type == ColorizationType.SPEED) {
						valList.add(analysis.pointAttributes.get(wptIdx).speed.toDouble())
					} else {
						valList.add(analysis.pointAttributes.get(wptIdx).elevation.toDouble())
					}
					wptIdx++
				}
			}
		}
		colorizationType = type
		latitudes = listToArray(latList)
		longitudes = listToArray(lonList)
		values = if (type == ColorizationType.SLOPE) {
			SlopeCalculator.calculateSlopesByElevations(
				latitudes,
				longitudes,
				listToArray(valList),
				SLOPE_RANGE.toDouble()
			)
		} else {
			listToArray(valList)
		}
		calculateMinMaxValue(analysis, maxProfileSpeed)
		if (fixedValues) {
			this.palette = if (isValidPalette(palette)) palette!! else getDefaultPalette(type)
		} else {
			val originalPalette = if (isValidPalette(palette)) {
				palette!!
			} else {
				getDefaultRelativePalette(type)
			}
			this.palette = ColorPalette(originalPalette, minValue, maxValue, type.bipolar)
		}
	}

	private fun isValidPalette(palette: ColorPalette?): Boolean {
		return palette != null && palette.colors.size >= 2
	}

	val result: List<RouteColorizationPoint>
		get() {
			val result = mutableListOf<RouteColorizationPoint>()
			for (i in latitudes.indices) {
				result.add(
					RouteColorizationPoint(
						i,
						latitudes[i], longitudes[i], values[i]
					)
				)
			}
			setColorsToPoints(result)
			return result
		}

	fun getSimplifiedResult(simplificationZoom: Int): List<RouteColorizationPoint> {
		val simplifiedResult = simplify(simplificationZoom)
		setColorsToPoints(simplifiedResult)
		return simplifiedResult
	}

	private fun setColorsToPoints(points: List<RouteColorizationPoint>) {
		for (point in points) {
			point.primaryColor = palette.getColorByValue(point.value)
		}
	}

	fun setPalette(palette: ColorPalette?) {
		if (palette != null) {
			this.palette = palette
		}
	}

	fun simplify(simplificationZoom: Int): List<RouteColorizationPoint> {
		var dataList = dataList
		if (dataList == null) {
			dataList = mutableListOf()
			for (i in latitudes.indices) {
				dataList.add(
					RouteColorizationPoint(
						i,
						latitudes[i], longitudes[i], values[i]
					)
				)
			}
			this.dataList = dataList
		}
		val points = mutableListOf<Node>()
		val result = mutableListOf<Node>()
		for (data in dataList) {
			points.add(Node(data.lat, data.lon, data.id))
		}
		val epsilon = 2.0.pow(DEFAULT_BASE - simplificationZoom)
		result.add(points[0])
		simplifyDouglasPeucker(
			points,
			0,
			points.size - 1,
			result,
			epsilon
		)
		val simplified= mutableListOf<RouteColorizationPoint>()
		for (i in 1 until result.size) {
			val prevId: Int = result[i - 1].id
			val currentId: Int = result[i].id
			val sublist: List<RouteColorizationPoint> = dataList.subList(prevId, currentId)
			simplified.addAll(getExtremums(sublist))
		}
		val lastSurvivedPoint = result[result.size - 1]
		simplified.add(dataList[lastSurvivedPoint.id])
		return simplified
	}

	private fun simplifyDouglasPeucker(
		points: List<Node>,
		start: Int,
		end: Int,
		survivedLPoints: MutableList<Node>,
		epsilon: Double
	) {
		var dmax = Double.NEGATIVE_INFINITY
		var index = -1
		val startPt = points[start]
		val endPt = points[end]
		for (i in start + 1 until end) {
			val pt = points[i]
			val d = KMapUtils.getOrthogonalDistance(
				pt.lat, pt.lon,
				startPt.lat, startPt.lon,
				endPt.lat, endPt.lon
			)
			if (d > dmax) {
				dmax = d
				index = i
			}
		}
		if (dmax > epsilon) {
			simplifyDouglasPeucker(points, start, index, survivedLPoints, epsilon)
			simplifyDouglasPeucker(points, index, end, survivedLPoints, epsilon)
		} else {
			survivedLPoints.add(points[end])
		}
	}

	private fun getExtremums(subDataList: List<RouteColorizationPoint>): List<RouteColorizationPoint> {
		if (subDataList.size <= 2) {
			return subDataList
		}
		val result = mutableListOf<RouteColorizationPoint>()
		var min: Double
		var max: Double
		max = subDataList[0].value
		min = max
		for (pt in subDataList) {
			if (min > pt.value) {
				min = pt.value
			}
			if (max < pt.value) {
				max = pt.value
			}
		}
		val diff = max - min
		result.add(subDataList[0])
		for (i in 1 until subDataList.size - 1) {
			val prev = subDataList[i - 1].value
			val current = subDataList[i].value
			val next = subDataList[i + 1].value
			val currentData = subDataList[i]
			if (current > prev && current > next || current < prev && current < next || current < prev && current == next || current == prev && current < next || current > prev && current == next || current == prev && current > next) {
				var prevInResult: RouteColorizationPoint
				if (result.size > 0) {
					prevInResult = result[0]
					if (prevInResult.value / diff > MIN_DIFFERENCE_SLOPE) {
						result.add(currentData)
					}
				} else result.add(currentData)
			}
		}
		result.add(subDataList[subDataList.size - 1])
		return result
	}

	private fun calculateMinMaxValue() {
		if (values.isEmpty()) return
		maxValue = Double.NaN
		minValue = maxValue
		for (value in values) {
			if ((maxValue.isNaN() || minValue.isNaN()) && !value.isNaN()) {
				minValue = value
				maxValue = minValue
			}
			if (minValue > value) minValue = value
			if (maxValue < value) maxValue = value
		}
	}

	private fun calculateMinMaxValue(
		analysis: GpxTrackAnalysis,
		maxProfileSpeed: Float
	) {
		calculateMinMaxValue()

		// set strict limitations for maxValue ONLY for linear (non-bipolar) types
		if (colorizationType?.bipolar == false) {
			maxValue = getMaxValue(colorizationType, analysis, minValue, maxProfileSpeed.toDouble())
		}
	}

	private fun listToArray(doubleList: List<Double>): DoubleArray {
		val result = DoubleArray(doubleList.size)
		for (i in doubleList.indices) {
			result[i] = doubleList[i]
		}
		return result
	}

	class RouteColorizationPoint(var id: Int, var lat: Double, var lon: Double, var value: Double) {
		var primaryColor = 0
		var secondaryColor = 0
	}

	private data class Node(val lat: Double, val lon: Double, val id: Int)

	companion object {
		private val LOG = LoggerFactory.getLogger("RouteColorize")

		var MAX_CORRECT_ELEVATION_DISTANCE = 100.0 // in meters
		var SLOPE_RANGE = 150 // 150 meters
		private const val DEFAULT_BASE = 17.2
		private const val MIN_DIFFERENCE_SLOPE = 0.05 // 5%

		fun getMinValue(
			type: ColorizationType?,
			analysis: GpxTrackAnalysis
		): Double {
			return when (type) {
				ColorizationType.SPEED -> 0.0
				ColorizationType.ELEVATION -> analysis.minElevation
				ColorizationType.SLOPE -> ColorPalette.SLOPE_MIN_VALUE
				else -> (-1.0)
			}
		}

		fun getMaxValue(
			type: ColorizationType?, analysis: GpxTrackAnalysis, minValue: Double,
			maxProfileSpeed: Double
		): Double {
			return when (type) {
				ColorizationType.SPEED -> max(analysis.maxSpeed.toDouble(), maxProfileSpeed)
				ColorizationType.ELEVATION -> max(analysis.maxElevation, minValue + 50)
				ColorizationType.SLOPE -> ColorPalette.SLOPE_MAX_VALUE
				else -> (-1.0)
			}
		}

		fun getDefaultPalette(colorizationType: ColorizationType): ColorPalette {
			return if (colorizationType == ColorizationType.SLOPE) {
				ColorPalette.SLOPE_PALETTE
			} else {
				ColorPalette.MIN_MAX_PALETTE
			}
		}

		fun getDefaultRelativePalette(colorizationType: ColorizationType): ColorPalette {
			return if (colorizationType.bipolar) {
				ColorPalette.BIPOLAR_MIN_MAX_PALETTE
			} else {
				ColorPalette.MIN_MAX_PALETTE
			}
		}
	}
}
